File: scripts/processIncidents.py
import sys, os
sys.stdout.reconfigure(encoding='utf-8')
import pandas as pd
import numpy as np
import json
from math import radians, cos, sin, asin, sqrt

INPUT_FILE  = os.path.join(os.path.dirname(__file__), '..', 'data', 'incidentes_ditra.xlsx')
OUTPUT_FILE = os.path.join(os.path.dirname(__file__), '..', 'src', 'data', 'hotspots_processed.json')

CORRIDOR_BOUNDS = {
    'C1': {'latMin': 4.70, 'latMax': 6.60, 'lngMin': -75.70, 'lngMax': -74.10},
    'C2': {'latMin': 2.10, 'latMax': 4.85, 'lngMin': -77.00, 'lngMax': -75.80},
    'C3': {'latMin': 4.15, 'latMax': 4.75, 'lngMin': -74.95, 'lngMax': -74.20},
    'C4': {'latMin': 4.28, 'latMax': 4.55, 'lngMin': -75.65, 'lngMax': -75.00},
    'C5': {'latMin': 3.95, 'latMax': 4.75, 'lngMin': -74.10, 'lngMax': -73.40},
    'C6': {'latMin': 4.70, 'latMax': 5.75, 'lngMin': -74.10, 'lngMax': -73.20},
    'C7': {'latMin': 10.55, 'latMax': 11.35, 'lngMin': -75.00, 'lngMax': -74.00},
}

NOMBRES = {
    'C1': 'Medellín–Honda–Bogotá', 'C2': 'Popayán–Cali–Cartago',
    'C3': 'Bogotá–Girardot', 'C4': 'Ibagué–La Línea–Cajamarca',
    'C5': 'Bogotá–Villavicencio', 'C6': 'Bogotá–Tunja',
    'C7': 'Santa Marta–Barranquilla',
}

PESO_SEV = {'MUERTO': 10, 'LESIONADO': 5, 'ACCIDENTADO': 2, 'SOLO DAÑOS': 1}
FACTOR_HORA = {'00:00 - 05:59': 1.80, '06:00 - 11:59': 1.10, '12:00 - 17:59': 1.00, '18:00 - 23:59': 1.40}
FACTOR_VEH = {'MOTOCICLETA': 1.50, 'CAMION': 1.30, 'BUS': 1.20, 'AUTOMOVIL': 1.00}
FACTOR_SS = 1.45

def haversine(lat1, lon1, lat2, lon2):
    lat1, lon1, lat2, lon2 = map(radians, [lat1, lon1, lat2, lon2])
    dlat, dlon = lat2-lat1, lon2-lon1
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    return 6371 * 2 * asin(sqrt(a))

def assign_corridor(lat, lng):
    for cid, b in CORRIDOR_BOUNDS.items():
        if b['latMin'] <= lat <= b['latMax'] and b['lngMin'] <= lng <= b['lngMax']:
            return cid
    return None

def calc_ira(row):
    evento = str(row.get('EVENTO', '')).upper().strip()
    sev = PESO_SEV.get(evento, 1)
    hora = FACTOR_HORA.get(row.get('Hechos_INTERVALOS_HORA', ''), 1.0)
    veh = FACTOR_VEH.get(str(row.get('CLASE_VEHICULO', '')).upper().strip(), 1.0)
    mes_num = int(row.get('Calend_MES_numero', 0) or 0)
    ss = FACTOR_SS if mes_num in (3, 4) else 1.0
    return sev * hora * veh * ss

def mode_val(series):
    if series.empty: return 'N/D'
    filtered = series.dropna()
    if filtered.empty: return 'N/D'
    return filtered.value_counts().index[0]

def cluster_hotspots(incidents_df, radius_km=0.5):
    if incidents_df.empty:
        return []

    # Convert to records for speed
    records = incidents_df.sort_values('ira', ascending=False).to_dict('records')

    # Grid-based spatial indexing (~0.005 degrees ≈ 500m)
    grid_size = 0.005
    grid = {}
    for r in records:
        gx = int(r['lat'] / grid_size)
        gy = int(r['lng'] / grid_size)
        key = (gx, gy)
        if key not in grid:
            grid[key] = []
        grid[key].append(r)

    assigned = set()
    clusters = []

    for rec in records:
        rid = rec['id']
        if rid in assigned:
            continue

        # Check only neighboring grid cells (3x3 around point)
        gx = int(rec['lat'] / grid_size)
        gy = int(rec['lng'] / grid_size)
        neighbors = []
        for dx in range(-1, 2):
            for dy in range(-1, 2):
                cell = grid.get((gx+dx, gy+dy), [])
                for other in cell:
                    if other['id'] in assigned:
                        continue
                    if haversine(rec['lat'], rec['lng'], other['lat'], other['lng']) <= radius_km:
                        neighbors.append(other)

        if len(neighbors) < 2:
            continue

        for m in neighbors:
            assigned.add(m['id'])

        total_ira = sum(m['ira'] for m in neighbors)
        cent_lat = sum(m['lat'] * m['ira'] for m in neighbors) / total_ira
        cent_lng = sum(m['lng'] * m['ira'] for m in neighbors) / total_ira

        # Frequency helpers
        def freq_top(key_name):
            f = {}
            for m in neighbors:
                v = m.get(key_name)
                if v and pd.notna(v):
                    f[v] = f.get(v, 0) + 1
            return sorted(f.items(), key=lambda x: -x[1])[0][0] if f else 'SIN DATOS'

        hip = freq_top('hipotesis')
        veh = freq_top('vehiculo')
        hora = freq_top('hora')

        muertos = sum(1 for m in neighbors if str(m.get('evento','')).upper().strip() == 'MUERTO')
        lesionados = sum(1 for m in neighbors if str(m.get('evento','')).upper().strip() == 'LESIONADO')
        sev_max = 'FATAL' if muertos > 0 else ('GRAVE' if lesionados > 0 else 'LEVE')
        ira_cluster = min(100, round(total_ira / len(neighbors) * 10))

        ss_count = sum(1 for m in neighbors if int(m.get('mesNum', 0) or 0) in (3, 4))
        factor_ss = round(ss_count / len(neighbors) * 100, 1)

        meses = list(set(m.get('mes') for m in neighbors if m.get('mes') and pd.notna(m.get('mes'))))
        years_raw = [m.get('año') for m in neighbors if m.get('año') and pd.notna(m.get('año'))]
        years = []
        for y in years_raw:
            try: years.append(int(y))
            except: pass
        years = sorted(set(years))

        clusters.append({
            'id': f'{rec["corridorId"]}-HS-{len(clusters)+1}',
            'corridorId': rec['corridorId'],
            'lat': round(cent_lat, 6),
            'lng': round(cent_lng, 6),
            'totalIncidentes': len(neighbors),
            'iraScore': int(ira_cluster),
            'severidadMax': sev_max,
            'hipotesisPrincipal': hip,
            'vehiculoPrincipal': veh,
            'horaCritica': hora,
            'muertos': muertos,
            'lesionados': lesionados,
            'meses': meses,
            'años': [str(y) for y in years],
            'factorSemSanta': float(factor_ss),
            'radioMetros': int(radius_km * 1000),
        })

    return clusters

def main():
    print('📂 Leyendo archivo Excel DITRA...')
    df = pd.read_excel(INPUT_FILE)
    print(f'✅ {len(df):,} registros cargados')

    # Detect year column — look for the AÑO column specifically (may have encoding issues)
    year_col = None
    for c in df.columns:
        if 'Calend' in c and ('AÑO' in c or 'A\u00d1O' in c or 'AÃ' in c or 'A?O' in c or c.endswith('O') and len(c) < 15):
            year_col = c
            break
    # Fallback: extract year from FECHA_HECHO
    if year_col is None or 'FECHA' in str(year_col):
        print('   Extrayendo año desde Calend_FECHA_HECHO...')
        df['_year_extracted'] = pd.to_datetime(df['Calend_FECHA_HECHO'], errors='coerce').dt.year
        year_col = '_year_extracted'
    print(f'   Columna de año: {year_col}')

    # Parse lat/lng
    df['lat'] = pd.to_numeric(df['LATITUD'], errors='coerce')
    df['lng'] = pd.to_numeric(df['LONGITUD'], errors='coerce')

    valid = df[(df['lat'].notna()) & (df['lng'].notna()) &
               (df['lat'] != 0) & (df['lng'] != 0) &
               (df['lat'] > -5) & (df['lat'] < 13) &
               (df['lng'] > -80) & (df['lng'] < -66)].copy()
    print(f'🗺️  {len(valid):,} registros con coordenadas válidas')

    # Assign corridors
    valid['corridorId'] = valid.apply(lambda r: assign_corridor(r['lat'], r['lng']), axis=1)
    corridored = valid[valid['corridorId'].notna()].copy()
    print(f'🛣️  {len(corridored):,} incidentes asignados a los 7 corredores')

    # Calculate IRA
    corridored['ira'] = corridored.apply(calc_ira, axis=1)
    corridored['id'] = corridored['Hechos_HECHOS_ID']
    corridored['hipotesis'] = corridored['HIPOTESIS_HECHOS_ACCI']
    corridored['evento'] = corridored['EVENTO']
    corridored['vehiculo'] = corridored['CLASE_VEHICULO']
    corridored['hora'] = corridored['Hechos_INTERVALOS_HORA']
    corridored['mes'] = corridored['Calend_MES']
    corridored['mesNum'] = corridored['Calend_MES_numero']
    corridored['año'] = corridored[year_col] if year_col else None
    corridored['municipio'] = corridored['Hechos_MUNICIPIO_HECHO']
    corridored['zona'] = corridored['Hechos_ZONA']

    # Process per corridor
    resumen = {}
    all_hotspots = []

    for cid in CORRIDOR_BOUNDS.keys():
        cinc = corridored[corridored['corridorId'] == cid]
        ira_total = cinc['ira'].sum()
        eventos_upper = cinc['evento'].str.upper().str.strip()

        resumen[cid] = {
            'corridorId': cid,
            'nombre': NOMBRES[cid],
            'totalIncidentes': int(len(cinc)),
            'muertos': int((eventos_upper == 'MUERTO').sum()),
            'lesionados': int((eventos_upper == 'LESIONADO').sum()),
            'iraPromedio': round(ira_total / len(cinc), 2) if len(cinc) > 0 else 0,
            'iraTotal': round(float(ira_total), 2),
            'pctSemSanta': round(pd.to_numeric(cinc['mesNum'], errors='coerce').isin([3,4]).sum() / max(len(cinc),1) * 100, 1),
            'vehiculoDominante': mode_val(cinc['vehiculo']),
            'horaCritica': mode_val(cinc['hora']),
            'hipotesisPrincipal': mode_val(cinc['hipotesis']),
        }

        # Cluster hotspots
        print(f'  {cid} ({NOMBRES[cid]}): {len(cinc)} incidentes...', end=' ', flush=True)
        hotspots = cluster_hotspots(cinc[['id','corridorId','lat','lng','ira','hipotesis','evento','vehiculo','hora','mes','mesNum','año']].copy(), 0.5)
        top15 = sorted(hotspots, key=lambda h: h['iraScore'], reverse=True)[:15]
        all_hotspots.extend(top15)
        print(f'→ {len(hotspots)} hotspots → top {len(top15)}')

    output = {
        'generado': pd.Timestamp.now().isoformat(),
        'totalRegistros': int(len(valid)),
        'totalCorredores': len(resumen),
        'totalHotspots': len(all_hotspots),
        'resumenPorCorredor': resumen,
        'hotspots': all_hotspots,
    }

    os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(output, f, indent=2, ensure_ascii=False)

    size_kb = os.path.getsize(OUTPUT_FILE) / 1024
    print(f'\n✅ Procesamiento completo → {OUTPUT_FILE}')
    print(f'   Total hotspots generados: {len(all_hotspots)}')
    print(f'   Tamaño del archivo: {size_kb:.0f} KB')

if __name__ == '__main__':
    main()

File: scripts/processIncidents.js
const XLSX = require('xlsx');
const fs   = require('fs');
const path = require('path');

const INPUT_FILE  = path.join(__dirname, '../data/incidentes_ditra.xlsx');
const OUTPUT_FILE = path.join(__dirname, '../src/data/hotspots_processed.json');

const CORRIDOR_BUFFER_KM = 2.0;

const CORRIDOR_BOUNDS = {
  C1: { latMin: 4.70, latMax: 6.60, lngMin: -75.70, lngMax: -74.10 },
  C2: { latMin: 2.10, latMax: 4.85, lngMin: -77.00, lngMax: -75.80 },
  C3: { latMin: 4.15, latMax: 4.75, lngMin: -74.95, lngMax: -74.20 },
  C4: { latMin: 4.28, latMax: 4.55, lngMin: -75.65, lngMax: -75.00 },
  C5: { latMin: 3.95, latMax: 4.75, lngMin: -74.10, lngMax: -73.40 },
  C6: { latMin: 4.70, latMax: 5.75, lngMin: -74.10, lngMax: -73.20 },
  C7: { latMin: 10.55, latMax: 11.35, lngMin: -75.00, lngMax: -74.00 },
};

const COLS = {
  ID:          'Hechos_HECHOS_ID',
  MUNICIPIO:   'Hechos_MUNICIPIO_HECHO',
  BARRIO:      'Hechos_BARRIOS_HECHO',
  CLASE:       'Hechos_SIEVI_CLASE_ACCIDENTE',
  ZONA:        'Hechos_ZONA',
  CUADRANTE:   'Hechos_CUADRANTE_VIAL',
  DIA_SEMANA:  'Calend_DIa_SeMaNa',
  MES:         'Calend_MES',
  MES_NUM:     'Calend_MES_numero',
  RUTA:        'Hechos_RUTA_HECHO',
  FECHA:       'Calend_FECHA_HECHO',
  HORA:        'Hechos_INTERVALOS_HORA',
  HIPOTESIS:   'HIPOTESIS_HECHOS_ACCI',
  GENERO:      'Person_GENERO',
  EDAD:        'Person_EDAD',
  GRUPO_EDAD:  'GRUPO_EDAD',
  EVENTO:      'EVENTO',
  PARTICIPANTE:'TIPO_PARTICIPANTE',
  VEHICULO:    'CLASE_VEHICULO',
  SERVICIO:    'TIPO_SERVICIO',
  LATITUD:     'LATITUD',
  LONGITUD:    'LONGITUD',
};

// Handle the year column with encoding issue
const COLS_YEAR = 'Calend_AÑO';
const COLS_YEAR_ALT = 'Calend_A\u00d1O';

const PESO_SEVERIDAD = {
  'MUERTO':     10,
  'LESIONADO':   5,
  'ACCIDENTADO': 2,
  'SOLO DAÑOS':  1,
};

const FACTOR_HORA = {
  '00:00 - 05:59': 1.80,
  '06:00 - 11:59': 1.10,
  '12:00 - 17:59': 1.00,
  '18:00 - 23:59': 1.40,
};

const FACTOR_VEHICULO = {
  'MOTOCICLETA': 1.50,
  'CAMION':      1.30,
  'BUS':         1.20,
  'AUTOMOVIL':   1.00,
};

const FACTOR_SEMANA_SANTA = 1.45;

function haversine(lat1, lon1, lat2, lon2) {
  const R = 6371;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = Math.sin(dLat/2)**2
            + Math.cos(lat1 * Math.PI/180) * Math.cos(lat2 * Math.PI/180)
            * Math.sin(dLon/2)**2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
}

function assignCorridor(lat, lng) {
  const hits = [];
  for (const [id, bounds] of Object.entries(CORRIDOR_BOUNDS)) {
    if (lat >= bounds.latMin && lat <= bounds.latMax &&
        lng >= bounds.lngMin && lng <= bounds.lngMax) {
      hits.push(id);
    }
  }
  return hits.length > 0 ? hits[0] : null;
}

function calcIRA(row) {
  const evento = (row[COLS.EVENTO] || '').toUpperCase().trim();
  const severidad  = PESO_SEVERIDAD[evento] ?? 1;
  const horaFactor = FACTOR_HORA[row[COLS.HORA]] ?? 1.0;
  const vehiculo = (row[COLS.VEHICULO] || '').toUpperCase().trim();
  const vehFactor  = FACTOR_VEHICULO[vehiculo] ?? 1.0;
  const mesNum     = parseInt(row[COLS.MES_NUM]) || 0;
  const esFestivo  = mesNum === 3 || mesNum === 4;
  const ssFactor   = esFestivo ? FACTOR_SEMANA_SANTA : 1.0;
  return severidad * horaFactor * vehFactor * ssFactor;
}

function clusterHotspots(incidents, radiusKm = 0.5) {
  const clusters = [];
  const assigned = new Set();
  const sorted = [...incidents].sort((a, b) => b.ira - a.ira);

  for (const inc of sorted) {
    if (assigned.has(inc.id)) continue;

    const members = sorted.filter(other => {
      if (assigned.has(other.id)) return false;
      return haversine(inc.lat, inc.lng, other.lat, other.lng) <= radiusKm;
    });

    if (members.length >= 2) {
      members.forEach(m => assigned.add(m.id));

      const totalIRA  = members.reduce((s, m) => s + m.ira, 0);
      const centLat   = members.reduce((s, m) => s + m.lat * m.ira, 0) / totalIRA;
      const centLng   = members.reduce((s, m) => s + m.lng * m.ira, 0) / totalIRA;

      const hipFreq  = {};
      members.forEach(m => { if (m.hipotesis) hipFreq[m.hipotesis] = (hipFreq[m.hipotesis] || 0) + 1; });
      const hipPrincipal = Object.entries(hipFreq).sort((a,b) => b[1]-a[1])[0]?.[0] ?? 'SIN DATOS';

      const vehFreq = {};
      members.forEach(m => { if (m.vehiculo) vehFreq[m.vehiculo] = (vehFreq[m.vehiculo] || 0) + 1; });
      const vehPrincipal = Object.entries(vehFreq).sort((a,b) => b[1]-a[1])[0]?.[0] ?? 'SIN DATOS';

      const horaFreq = {};
      members.forEach(m => { if (m.hora) horaFreq[m.hora] = (horaFreq[m.hora] || 0) + 1; });
      const horaCritica = Object.entries(horaFreq).sort((a,b) => b[1]-a[1])[0]?.[0] ?? 'SIN DATOS';

      const hasMuertos    = members.some(m => (m.evento || '').toUpperCase() === 'MUERTO');
      const hasLesionados = members.some(m => (m.evento || '').toUpperCase() === 'LESIONADO');
      const severidadMax  = hasMuertos ? 'FATAL' : hasLesionados ? 'GRAVE' : 'LEVE';
      const iraCluster    = Math.min(100, Math.round(totalIRA / members.length * 10));

      clusters.push({
        id:              `HS-${clusters.length + 1}`,
        corridorId:      members[0].corridorId,
        lat:             parseFloat(centLat.toFixed(6)),
        lng:             parseFloat(centLng.toFixed(6)),
        totalIncidentes: members.length,
        iraScore:        iraCluster,
        severidadMax,
        hipotesisPrincipal: hipPrincipal,
        vehiculoPrincipal:  vehPrincipal,
        horaCritica,
        muertos:    members.filter(m => (m.evento || '').toUpperCase() === 'MUERTO').length,
        lesionados: members.filter(m => (m.evento || '').toUpperCase() === 'LESIONADO').length,
        meses: [...new Set(members.map(m => m.mes))].filter(Boolean),
        años:  [...new Set(members.map(m => m.año))].filter(Boolean).sort(),
        factorSemSanta: parseFloat(
          (members.filter(m => [3, 4].includes(parseInt(m.mesNum))).length / members.length * 100).toFixed(1)
        ),
        radioMetros: Math.round(radiusKm * 1000),
      });
    }
  }
  return clusters;
}

function processIncidents() {
  console.log('📂 Leyendo archivo Excel DITRA...');
  const wb   = XLSX.readFile(INPUT_FILE, { cellText: false, cellDates: true });
  const ws   = wb.Sheets[wb.SheetNames[0]];
  const rows = XLSX.utils.sheet_to_json(ws, { defval: null });
  console.log(`✅ ${rows.length.toLocaleString()} registros cargados`);

  // Detect year column name
  const sampleRow = rows[0] || {};
  const yearCol = Object.keys(sampleRow).find(k => k.includes('AÑO') || k.includes('A\u00d1O') || k.includes('AÃ') || k.includes('A?O'));
  console.log(`   Columna de año detectada: "${yearCol}"`);

  const valid = rows.filter(r => {
    const lat = parseFloat(r[COLS.LATITUD]);
    const lng = parseFloat(r[COLS.LONGITUD]);
    return !isNaN(lat) && !isNaN(lng) && lat !== 0 && lng !== 0
           && lat > -5 && lat < 13 && lng > -80 && lng < -66;
  });
  console.log(`🗺️  ${valid.length.toLocaleString()} registros con coordenadas válidas`);

  const incidents = [];
  for (const row of valid) {
    const lat = parseFloat(row[COLS.LATITUD]);
    const lng = parseFloat(row[COLS.LONGITUD]);
    const corridorId = assignCorridor(lat, lng);
    if (!corridorId) continue;

    incidents.push({
      id:          row[COLS.ID],
      corridorId,
      lat, lng,
      municipio:   row[COLS.MUNICIPIO],
      zona:        row[COLS.ZONA],
      ruta:        row[COLS.RUTA],
      hipotesis:   row[COLS.HIPOTESIS],
      evento:      row[COLS.EVENTO],
      vehiculo:    row[COLS.VEHICULO],
      participante:row[COLS.PARTICIPANTE],
      hora:        row[COLS.HORA],
      diaSemana:   row[COLS.DIA_SEMANA],
      mes:         row[COLS.MES],
      mesNum:      row[COLS.MES_NUM],
      año:         yearCol ? row[yearCol] : null,
      ira:         calcIRA(row),
    });
  }

  console.log(`🛣️  ${incidents.length.toLocaleString()} incidentes asignados a los 7 corredores`);

  const NOMBRES_CORREDOR = {
    C1: 'Medellín–Honda–Bogotá', C2: 'Popayán–Cali–Cartago',
    C3: 'Bogotá–Girardot',       C4: 'Ibagué–La Línea–Cajamarca',
    C5: 'Bogotá–Villavicencio',  C6: 'Bogotá–Tunja',
    C7: 'Santa Marta–Barranquilla',
  };

  const resumenPorCorredor = {};
  for (const cid of Object.keys(CORRIDOR_BOUNDS)) {
    const cinc = incidents.filter(i => i.corridorId === cid);
    const iraTotal = cinc.reduce((s, i) => s + i.ira, 0);

    resumenPorCorredor[cid] = {
      corridorId:    cid,
      nombre:        NOMBRES_CORREDOR[cid],
      totalIncidentes: cinc.length,
      muertos:       cinc.filter(i => (i.evento || '').toUpperCase() === 'MUERTO').length,
      lesionados:    cinc.filter(i => (i.evento || '').toUpperCase() === 'LESIONADO').length,
      iraPromedio:   cinc.length > 0 ? parseFloat((iraTotal / cinc.length).toFixed(2)) : 0,
      iraTotal:      parseFloat(iraTotal.toFixed(2)),
      pctSemSanta:   cinc.length > 0
        ? parseFloat((cinc.filter(i => [3,4].includes(parseInt(i.mesNum))).length / cinc.length * 100).toFixed(1))
        : 0,
      vehiculoDominante: (() => {
        const f = {}; cinc.forEach(i => { if(i.vehiculo) f[i.vehiculo] = (f[i.vehiculo]||0)+1; });
        return Object.entries(f).sort((a,b)=>b[1]-a[1])[0]?.[0] ?? 'N/D';
      })(),
      horaCritica: (() => {
        const f = {}; cinc.forEach(i => { if(i.hora) f[i.hora] = (f[i.hora]||0)+1; });
        return Object.entries(f).sort((a,b)=>b[1]-a[1])[0]?.[0] ?? 'N/D';
      })(),
      hipotesisPrincipal: (() => {
        const f = {}; cinc.filter(i=>i.hipotesis).forEach(i => { f[i.hipotesis] = (f[i.hipotesis]||0)+1; });
        return Object.entries(f).sort((a,b)=>b[1]-a[1])[0]?.[0] ?? 'SIN DATOS';
      })(),
    };
  }

  const allHotspots = [];
  for (const cid of Object.keys(CORRIDOR_BOUNDS)) {
    const cinc = incidents.filter(i => i.corridorId === cid);
    const hotspots = clusterHotspots(cinc, 0.5);
    const top15 = hotspots.sort((a, b) => b.iraScore - a.iraScore).slice(0, 15);
    allHotspots.push(...top15);
    console.log(`  ${cid} (${NOMBRES_CORREDOR[cid]}): ${cinc.length} incidentes → ${hotspots.length} hotspots → top ${top15.length}`);
  }

  const output = {
    generado:        new Date().toISOString(),
    totalRegistros:  valid.length,
    totalCorredores: Object.keys(resumenPorCorredor).length,
    totalHotspots:   allHotspots.length,
    resumenPorCorredor,
    hotspots: allHotspots,
  };

  fs.mkdirSync(path.dirname(OUTPUT_FILE), { recursive: true });
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(output, null, 2));
  console.log(`\n✅ Procesamiento completo → ${OUTPUT_FILE}`);
  console.log(`   Total hotspots generados: ${allHotspots.length}`);
  console.log(`   Tamaño del archivo: ${(fs.statSync(OUTPUT_FILE).size / 1024).toFixed(0)} KB`);
}

processIncidents();
